File: src/lib/alife/network.ts
// ═══════════════════════════════════════════════════════════════════
//  network.ts — Compiled feedforward neural network from a Genome
//
//  Base: Python network.py (Kahn's algorithm with min-heap)
//  Invariant n°4: topological sort computed ONCE at construction.
//  Hidden → tanh activation, Output → linear.
//
//  poc2.6 addition: Reward-modulated Hebbian plasticity.
//    - Optional HebbianConfig at construction (opt-in, zero cost when off)
//    - activate() stores activations for apply_hebbian()
//    - apply_hebbian() does V2 contrastive update every tick
//    - Non-Lamarckian: never writes back to genome
// ═══════════════════════════════════════════════════════════════════

import { NodeType, type Genome } from './genome';
import type { NetworkConfig, HebbianConfig } from './config';
import type { SeededRNG } from './genome';

const ACTIVATIONS: Record<string, (x: number) => number> = {
  tanh: Math.tanh,
};

function getActivation(name: string): (x: number) => number {
  const fn = ACTIVATIONS[name];
  if (!fn) throw new Error(`unknown activation '${name}'`);
  return fn;
}

type InEdge = [srcId: number, weight: number];

export class NeuralNetwork {
  private _activation: (x: number) => number;
  private _evalOrder: number[];
  private _nodeTypes: Map<number, string>;
  private _incoming: Map<number, InEdge[]>;
  private _biasIds: number[];
  readonly inputIds: number[];
  readonly outputIds: number[];

  // poc2.6: Hebbian plasticity state (allocated only when enabled)
  private _hebbian: HebbianConfig | null;
  private _lastValues: Map<number, number> | null = null;
  private _trace: Map<number, number[]> | null = null;
  private _rewardBaseline = 0.0;

  constructor(genome: Genome, _config: NetworkConfig, hebbian?: HebbianConfig) {
    this._activation = getActivation(_config.activation);

    // Plasticity is opt-in and costs the legacy path nothing.
    this._hebbian = (hebbian !== undefined && hebbian.enabled) ? hebbian : null;

    // Node type lookup
    this._nodeTypes = new Map();
    for (const n of genome.nodes) {
      this._nodeTypes.set(n.id, n.type);
    }

    this.inputIds = genome.inputIds;
    this.outputIds = genome.outputIds;
    this._biasIds = genome.biasIds;

    // Build incoming adjacency (enabled connections only)
    this._incoming = new Map();
    for (const n of genome.nodes) {
      this._incoming.set(n.id, []);
    }
    for (const conn of genome.connections) {
      if (conn.enabled) {
        this._incoming.get(conn.out_node)!.push([conn.in_node, conn.weight]);
      }
    }

    this._evalOrder = this._topoSort(genome);

    // Safety net: auto-repair cycles
    if (this._evalOrder.length !== genome.nodes.length) {
      this._repairCycle(genome);
    }

    // poc2.6: Allocate eligibility traces (parallel to _incoming)
    if (this._hebbian !== null) {
      this._trace = new Map();
      for (const [nid, srcs] of this._incoming) {
        this._trace.set(nid, new Array(srcs.length).fill(0.0));
      }
    }
  }

  /** Kahn's algorithm with min-heap. */
  private _topoSort(genome: Genome): number[] {
    const successors = new Map<number, number[]>();
    for (const n of genome.nodes) successors.set(n.id, []);
    for (const conn of genome.connections) {
      if (conn.enabled) {
        successors.get(conn.in_node)!.push(conn.out_node);
      }
    }

    const inDeg = new Map<number, number>();
    for (const n of genome.nodes) inDeg.set(n.id, 0);
    for (const conn of genome.connections) {
      if (conn.enabled) {
        inDeg.set(conn.out_node, (inDeg.get(conn.out_node) || 0) + 1);
      }
    }

    const heap = new MinHeap<number>();
    for (const [nid, deg] of inDeg) {
      if (deg === 0) heap.push(nid);
    }

    const order: number[] = [];
    while (heap.size > 0) {
      const nid = heap.pop()!;
      order.push(nid);
      for (const dst of successors.get(nid) || []) {
        const newDeg = (inDeg.get(dst) || 1) - 1;
        inDeg.set(dst, newDeg);
        if (newDeg === 0) heap.push(dst);
      }
    }
    return order;
  }

  private _repairCycle(genome: Genome): void {
    const unreachable = new Set<number>();
    for (const n of genome.nodes) {
      if (!this._evalOrder.includes(n.id)) unreachable.add(n.id);
    }
    for (const conn of genome.connections) {
      if (conn.enabled && unreachable.has(conn.out_node)) {
        conn.enabled = false;
      }
    }
    this._rebuildIncoming(genome);
    this._evalOrder = this._topoSort(genome);
    if (this._evalOrder.length === genome.nodes.length) return;
    for (const conn of genome.connections) {
      if (conn.enabled) {
        const srcType = this._nodeTypes.get(conn.in_node);
        const dstType = this._nodeTypes.get(conn.out_node);
        if (srcType === NodeType.HIDDEN || dstType === NodeType.HIDDEN) {
          conn.enabled = false;
        }
      }
    }
    this._rebuildIncoming(genome);
    this._evalOrder = this._topoSort(genome);
  }

  private _rebuildIncoming(genome: Genome): void {
    this._incoming = new Map();
    for (const n of genome.nodes) {
      this._incoming.set(n.id, []);
    }
    for (const conn of genome.connections) {
      if (conn.enabled) {
        this._incoming.get(conn.out_node)!.push([conn.in_node, conn.weight]);
      }
    }
  }

  /**
   * Forward pass. Returns raw (linear) output values.
   * Under Hebbian, stores activations for apply_hebbian().
   */
  activate(inputs: number[], rng?: SeededRNG): [number, number] {
    if (inputs.length !== this.inputIds.length) {
      throw new Error(`expected ${this.inputIds.length} inputs, got ${inputs.length}`);
    }

    const values = new Map<number, number>();
    for (let i = 0; i < this.inputIds.length; i++) {
      values.set(this.inputIds[i], inputs[i]);
    }
    for (const bid of this._biasIds) {
      values.set(bid, 1.0);
    }

    for (const nid of this._evalOrder) {
      const ntype = this._nodeTypes.get(nid);
      if (ntype === NodeType.INPUT || ntype === NodeType.BIAS) continue;

      const edges = this._incoming.get(nid)!;
      let total = 0;
      for (const [src, w] of edges) {
        total += w * (values.get(src) || 0);
      }

      if (ntype === NodeType.OUTPUT) {
        values.set(nid, total);
      } else {
        values.set(nid, this._activation(total));
      }
    }

    // poc2.6: Store activations for Hebbian update
    if (this._hebbian !== null) {
      this._lastValues = values;
    }

    let out0 = values.get(this.outputIds[0]) || 0;
    let out1 = values.get(this.outputIds[1]) || 0;

    if (out0 === 0 && out1 === 0 && rng) {
      out0 = rng.gauss(0, 0.1);
      out1 = rng.gauss(0, 0.1);
    }

    return [out0, out1];
  }

  /**
   * Advance eligibility traces and apply one contrastive Hebbian update.
   *
   * Called once per tick per agent (V2 — every tick, not just capture ticks):
   *   e  <- eligibility_decay * e + x * y
   *   dw = learning_rate * (reward - baseline) * e
   *
   * The baseline is read BEFORE folding in this tick's reward (genuine surprise signal).
   * Clamped to ±hebbian.weight_max. No-op when plasticity is off.
   *
   * CRITICAL: NOT called from activate(). Several callers activate a network
   * purely to measure it (steer_score, novelty descriptor, HyperNEAT CPPN queries)
   * and observing an agent must never modify its brain.
   */
  applyHebbian(reward: number): void {
    if (this._hebbian === null || this._lastValues === null) return;

    const values = this._lastValues;
    const cfg = this._hebbian;
    const decay = cfg.eligibility_decay;
    const limit = cfg.weight_max;

    // Contrast against expected BEFORE this tick's outcome
    const step = cfg.learning_rate * (reward - this._rewardBaseline);
    this._rewardBaseline += cfg.baseline_rate * (reward - this._rewardBaseline);

    for (const [nid, srcs] of this._incoming) {
      if (srcs.length === 0) continue;
      let post = values.get(nid) ?? 0;
      // Output nodes are linear, but Agent applies tanh to use them.
      // Learning on the emitted value keeps x and y in (-1, 1).
      if (this._nodeTypes.get(nid) === NodeType.OUTPUT) {
        post = Math.tanh(post);
      }
      const trace = this._trace!.get(nid)!;
      for (let pos = 0; pos < srcs.length; pos++) {
        const pre = values.get(srcs[pos][0]) ?? 0;
        const e = decay * trace[pos] + pre * post;
        trace[pos] = e;
        srcs[pos][1] = Math.max(-limit, Math.min(limit, srcs[pos][1] + step * e));
      }
    }
  }
}

// ── Min-heap for deterministic Kahn's sort ──────────────────

class MinHeap<T extends number> {
  private _data: T[] = [];
  get size(): number { return this._data.length; }
  push(val: T): void { this._data.push(val); this._bubbleUp(this._data.length - 1); }
  pop(): T | undefined {
    const top = this._data[0];
    const last = this._data.pop();
    if (this._data.length > 0 && last !== undefined) {
      this._data[0] = last; this._sinkDown(0);
    }
    return top;
  }
  private _bubbleUp(i: number): void {
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this._data[i] < this._data[parent]) {
        [this._data[i], this._data[parent]] = [this._data[parent], this._data[i]];
        i = parent;
      } else break;
    }
  }
  private _sinkDown(i: number): void {
    const n = this._data.length;
    while (true) {
      let smallest = i;
      const left = 2 * i + 1;
      const right = 2 * i + 2;
      if (left < n && this._data[left] < this._data[smallest]) smallest = left;
      if (right < n && this._data[right] < this._data[smallest]) smallest = right;
      if (smallest !== i) {
        [this._data[i], this._data[smallest]] = [this._data[smallest], this._data[i]];
        i = smallest;
      } else break;
    }
  }
}

File: src/network.py
"""Neural network evaluation built from a Genome (NEAT-style, feedforward only).

Invariants honoured here:
- Topological sort computed ONCE at __init__, never recomputed (CLAUDE.md n°4).
- Hidden nodes use the configured activation (tanh); output nodes are linear.
- Disabled connections are ignored everywhere (genome.py may leave them in place
  after add_node splits an edge).

Under ``hebbian.enabled`` the compiled weights are no longer frozen for the
agent's life: ``apply_hebbian`` adjusts them from the last forward pass (see
that method). The TOPOLOGY still never changes, so invariant n°4's cached
topological sort stays valid — only the weight copies move, and only in the
compiled network, never in the genome.
"""

from __future__ import annotations

import heapq
import math
from typing import Callable, Sequence

from src.config import HebbianConfig, NetworkConfig
from src.genome import BIAS, INPUT, OUTPUT, Genome

_ACTIVATIONS: dict[str, Callable[[float], float]] = {
    "tanh": math.tanh,
}


def _get_activation(name: str) -> Callable[[float], float]:
    try:
        return _ACTIVATIONS[name]
    except KeyError as exc:
        raise ValueError(
            f"unknown activation '{name}'; available: {sorted(_ACTIVATIONS)}"
        ) from exc


class NeuralNetwork:
    """Compiled, executable view of a Genome.

    Build once per agent at birth; call activate() every tick.
    """

    def __init__(
        self,
        genome: Genome,
        config: NetworkConfig,
        hebbian: HebbianConfig | None = None,
    ) -> None:
        self._activation = _get_activation(config.activation)
        # Plasticity is opt-in and costs the legacy path nothing: with it off,
        # activate() never records activations and the weights stay exactly the
        # frozen copies taken from the genome below.
        self._hebbian = hebbian if hebbian is not None and hebbian.enabled else None
        # Activations of the most recent activate(), kept only under plasticity
        # so apply_hebbian() can use the SAME tick's pre/postsynaptic values.
        self._last_values: dict[int, float] | None = None

        node_type: dict[int, str] = {n.node_id: n.node_type for n in genome.nodes}

        self.input_ids: list[int] = sorted(
            n.node_id for n in genome.nodes if n.node_type == INPUT
        )
        self.output_ids: list[int] = sorted(
            n.node_id for n in genome.nodes if n.node_type == OUTPUT
        )
        self._bias_ids: list[int] = [
            n.node_id for n in genome.nodes if n.node_type == BIAS
        ]

        # Build adjacency: node_id → [(src_id, weight), ...]  (enabled only)
        incoming: dict[int, list[tuple[int, float]]] = {
            n.node_id: [] for n in genome.nodes
        }
        for conn in genome.connections:
            if conn.enabled:
                incoming[conn.out_node].append((conn.in_node, conn.weight))

        # Kahn's algorithm with a min-heap — ascending node_id order for determinism
        in_deg: dict[int, int] = {nid: len(srcs) for nid, srcs in incoming.items()}
        heap: list[int] = [nid for nid, d in in_deg.items() if d == 0]
        heapq.heapify(heap)
        # successors: node → [out_node, ...]  (enabled connections only)
        successors: dict[int, list[int]] = {n.node_id: [] for n in genome.nodes}
        for conn in genome.connections:
            if conn.enabled:
                successors[conn.in_node].append(conn.out_node)
        eval_order: list[int] = []
        while heap:
            nid = heapq.heappop(heap)
            eval_order.append(nid)
            for dst in successors[nid]:
                in_deg[dst] -= 1
                if in_deg[dst] == 0:
                    heapq.heappush(heap, dst)

        if len(eval_order) != len(genome.nodes):
            raise ValueError(
                "genome contains a cycle — feedforward invariant violated "
                f"(processed {len(eval_order)}/{len(genome.nodes)} nodes)"
            )

        self._eval_order: list[int] = eval_order
        self._node_type: dict[int, str] = node_type
        self._incoming: dict[int, list[tuple[int, float]]] = incoming

    def activate(self, inputs: Sequence[float]) -> tuple[float, ...]:
        """Run a forward pass; return one raw value per output node, in id order.

        For an agent's own network this is always ``(vx_raw, vy_raw)`` (2
        outputs, per invariant n°5) — but the class is otherwise agnostic to
        output count, which a HyperNEAT CPPN (1 output: a queried connection
        weight, see ``src.hyperneat``) relies on.
        """
        if len(inputs) != len(self.input_ids):
            raise ValueError(
                f"expected {len(self.input_ids)} inputs, got {len(inputs)}"
            )

        values: dict[int, float] = {}
        for pos, nid in enumerate(self.input_ids):
            values[nid] = float(inputs[pos])
        for nid in self._bias_ids:
            values[nid] = 1.0  # always-on, never read from the sensor vector

        for nid in self._eval_order:
            ntype = self._node_type[nid]
            if ntype in (INPUT, BIAS):
                continue  # already set above
            total = sum(w * values[src] for src, w in self._incoming[nid])
            if ntype == OUTPUT:
                values[nid] = total  # linear output
            else:  # HIDDEN
                values[nid] = self._activation(total)

        if self._hebbian is not None:
            self._last_values = values

        return tuple(values[nid] for nid in self.output_ids)

    def apply_hebbian(self, reward: float) -> None:
        """Reward-modulated Hebbian update on the last forward pass.

        ``dw = learning_rate * reward * x * y`` for every enabled connection,
        clamped to ``+/- hebbian.weight_max``. No-op when plasticity is off,
        when ``reward`` is zero (the rule is reward-gated), or before the first
        :meth:`activate`.

        Deliberately NOT called from :meth:`activate`. Several callers activate
        a network purely to measure it — ``diagnostics.steer_score``, the
        novelty ``behavior_descriptor``, the HyperNEAT CPPN queries — and
        observing an agent must never modify its brain. Only ``Agent.eat``,
        which knows the tick's real reward, drives learning.

        Mutates the compiled network's own weight copies, never the genome:
        learning is non-Lamarckian, children inherit the innate wiring.
        """
        if self._hebbian is None or reward == 0.0 or self._last_values is None:
            return
        values = self._last_values
        step = self._hebbian.learning_rate * reward
        limit = self._hebbian.weight_max
        for nid, srcs in self._incoming.items():
            if not srcs:
                continue
            post = values[nid]
            if self._node_type[nid] == OUTPUT:
                # Output nodes are linear (unbounded), but what the node
                # actually EMITS is tanh(total) — Agent.decide applies it before
                # using the value. Learning on the raw sum would let one
                # saturated output dominate every update and would make
                # learning_rate uninterpretable; on the emitted value both
                # x and y are in (-1, 1), so one apple moves a weight by at
                # most learning_rate.
                post = math.tanh(post)
            self._incoming[nid] = [
                (src, max(-limit, min(limit, w + step * values[src] * post)))
                for src, w in srcs
            ]
